osu.Server.Queues.ScoreStatisticsProcessor/Commands/Performance/Scores/UpdateAllScoresCommand.cs

```diff
@@ -0,0 +1,120 @@
+        {
+            using var db = DatabaseAccess.GetConnection();
+
+            ulong currentScoreId = From;
+            ulong? lastScoreId = await db.QuerySingleAsync<ulong>("SELECT MAX(id) FROM scores");
+
+            ulong processedCount = 0;
+            ulong changedPp = 0;
+            double rate = 0;
+            Stopwatch sw = new Stopwatch();
+
+            var scoresQuery = db.Query<SoloScore>("SELECT * FROM scores WHERE `id` > @ScoreId AND `id` <= @LastScoreId ORDER BY `id`", new
+            {
+                ScoreId = currentScoreId,
+                LastScoreId = lastScoreId,
+            }, buffered: false);
+
+            using var scoresEnum = scoresQuery.GetEnumerator();
+
+            Console.WriteLine($"Processing all scores up to {lastScoreId}, starting from {currentScoreId}");
+
+            Task<List<SoloScore>> nextScores = getNextScores();
+
+            while (!cancellationToken.IsCancellationRequested)
+            {
+                if (CheckSlaveLatency)
+                    await SlaveLatencyChecker.CheckSlaveLatency(db, cancellationToken);
+
+                sw.Restart();
+
+                var scores = await nextScores;
+                nextScores = getNextScores();
+
+                if (scores.Count == 0)
+                    break;
+
+                await Task.WhenAll(Partitioner.Create(scores).GetPartitions(Threads).Select(async partition =>
+                {
+                    using (var connection = DatabaseAccess.GetConnection())
+                    using (var transaction = await connection.BeginTransactionAsync(IsolationLevel.ReadUncommitted, cancellationToken))
+                    using (partition)
+                    {
+                        while (partition.MoveNext())
+                        {
+                            if (cancellationToken.IsCancellationRequested)
+                                return;
+
+                            bool changed = await ScoreProcessor.ProcessScoreAsync(partition.Current, connection, transaction);
+
+                            if (changed)
+                                Interlocked.Increment(ref changedPp);
+                        }
+
+                        await transaction.CommitAsync(cancellationToken);
+                    }
+                }));
+
+                Interlocked.Add(ref processedCount, (ulong)scores.Count);
+
+                currentScoreId = scores.Last().id;
+
+                if (rate == 0)
+                    rate = ((double)scores.Count / sw.ElapsedMilliseconds * 1000);
+                else
+                    rate = rate * 0.95 + 0.05 * ((double)scores.Count / sw.ElapsedMilliseconds * 1000);
+
+                Console.WriteLine(ScoreProcessor.BeatmapStore?.GetCacheStats());
+                Console.WriteLine($"id: {currentScoreId:N0} changed: {changedPp:N0} ({processedCount:N0} of {lastScoreId:N0} {(float)processedCount / lastScoreId:P1}) {rate:N0}/s");

```

Comment:
Nit: `{count} of {lastScoreId}` is not really correct but I guess it's hard to do much better. Maybe `of {lastScoreId - From}` could be tried but /shrug.

Reply:
Yeah, it's kinda best-effort. I'll make it read better.

Reply:
4fdf3a9e9bb7fa8fe36bae993385a6d33d097163

![JetBrains Rider-EAP 2024-10-21 at 10 55 49](https://github.com/user-attachments/assets/09466a77-a2f1-4089-9a64-b753e2b30340)

No real point showing the max ID so I just removed that and modified the % part as you proposed.